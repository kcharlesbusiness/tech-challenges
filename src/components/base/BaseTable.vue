<script setup lang="ts">
import { onMounted } from 'vue';
// composables
import type { TableData } from '../../composables/tableComposable';
import { useSearch } from '../../composables/tableComposable';
// services
import { debounce } from '../../services/eventsService';

export interface BaseTableInterface {
  triggerSearchTable: () => void;
  handleSearch: (...args: any) => void;
}

const props = defineProps<{
  title: string,
  data: TableData,
  searchable?: Boolean,
}>();

const { state, searchTable } = useSearch();

const triggerSearchTable = () => {
  searchTable(state.default as unknown as TableData, state.searchTerm);
}
const handleSearch = debounce(triggerSearchTable);

onMounted(() => {
  state.loading = true;
  state.data = props.data;
  state.default = props.data;
  state.loading = false;
});

defineExpose<BaseTableInterface>({
  triggerSearchTable,
  handleSearch,
});
</script>

<template>
  <section class="v-base-table">
    <div class="container">
      <div class="v-base-table__table">
        <h2 class="">{{ props.title }}</h2>
        <section>
          <template v-if="props.searchable">
            <div>
              <input
                  type="search"
                  name="search"
                  v-model="state.searchTerm"
                  placeholder="search..."
                  @keyup.prevent="handleSearch"
              />
            </div>
          </template>
          <template v-if="data.length">
            <o-table
                :data="state.data"
                :bordered="true"
                :striped="false"
                :narrowed="false"
                :hoverable="true"
                :loading="state.loading"
                :focusable="false"
                :mobile-cards="false"
                :paginated="true"
                :per-page="24"
                :current-page.sync="1"
                :pagination-simple="true"
                :default-sort="Object.keys(data[0])[0]"
                sort-icon="arrow-up"
                sort-icon-size="small"
                aria-next-label="Next page"
                aria-previous-label="Previous page"
                aria-page-label="Page"
                aria-current-label="Current page"
            >
              <slot name="columns" :props="props" />
            </o-table>
          </template>
        </section>
      </div>
    </div>
  </section>
</template>

<style lang="scss">
.v-base-table {
  .o-table__wrapper {
    overflow: scroll;
  }

  input {
    padding: 0.5rem 1rem;
    margin-bottom: 1rem;
  }

  &__table {
    margin-top: 4rem;
    text-align: left;

    h2 {
      margin-bottom: 2rem;
    }
  }
}
</style>
