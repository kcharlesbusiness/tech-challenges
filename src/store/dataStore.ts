import {defineStore, StoreDefinition} from 'pinia';
import { DataTypeConstants } from '../constants/DataTypeConstants';
import BidsInterface from "../interfaces/BidsInterface";
import BuyerInterface, { TopBuyerInterface } from '../interfaces/BuyerInterface';
import CompetitionInterface, { ClosedCompetitionInterface } from '../interfaces/CompetitionInterface';
import SellerInterface from '../interfaces/SellerInterface';
import { getBidsList, getBuyerList, getCompetitionList, getSellerList } from '../services/dataService';
import Bid from '../models/Bid';
import Buyer from '../models/Buyer';
import Competition from '../models/Competition';
import { Ref } from 'vue';
import Seller from '../models/Seller';
import { CompetitionConstants } from '../constants/CompetitionConstants';

export type DataState = {
  bids: Array<BidsInterface>;
  buyers: Array<BuyerInterface>;
  competitions: Array<CompetitionInterface>;
  sellers: Array<SellerInterface>;
}

export const useDataStore = defineStore({
  id: 'dataStore',
  state: () => ({
    bids: [],
    buyers: [],
    competitions: [],
    sellers: [],
  }) as unknown as DataState,

  actions: {
    hydrateStore() {
      // @ts-ignore
      this.bids = getBidsList().map((item: BidsInterface) => {
        return new Bid(item);
      }) as Array<BidsInterface>;

      this.buyers = getBuyerList().map((item: BuyerInterface) => {
        return new Buyer(item);
      }) as Array<BuyerInterface>;

      // @ts-ignore
      this.competitions = getCompetitionList().map((item: CompetitionInterface) => {
        return new Competition(item);
      }) as Array<CompetitionInterface>;

      this.sellers = getSellerList().map((item: SellerInterface) => {
        return new Seller(item);
      }) as Array<SellerInterface>;
    },
    findById<T extends { id: string }>(type: DataTypeConstants, id: string): T | null {
      const matched: Array<T> = (this[`${type}s`] as Array<T>)
        .filter((item: T) => item.id === id) ?? null;

      return matched?.length ? matched[0] as T : null;
    },
    findBidById(id: string): BidsInterface | null {
      return this.findById<BidsInterface>(DataTypeConstants.BID, id) as unknown as BidsInterface;
    },
    findBuyerById(id: string): BuyerInterface | null {
      return this.findById(DataTypeConstants.BUYER, id) as unknown as BuyerInterface;
    },
    findCompetitionById(id: string): CompetitionInterface | null {
      return this.findById(DataTypeConstants.COMPETITION, id) as unknown as CompetitionInterface;
    },
    findSellerById(id: string): SellerInterface | null {
      return this.findById(DataTypeConstants.SELLER, id) as unknown as SellerInterface;
    },
    filterBy<T extends {}>(type: DataTypeConstants, callback: (item: T) => boolean): Array<T> {
      return (this[`${type}s`] as Array<T>).filter((item: T) => callback(item as T)) ?? null;
    },
  },

  getters: {
    getSuccessfulBids: (state: DataState): Array<BidsInterface> => {
      return state.bids.filter((bid: BidsInterface) => {
        return bid.isSuccessful;
      });
    },
    getClosedCompetitions: (state: DataState): Array<ClosedCompetitionInterface> => {
      let query: Array<ClosedCompetitionInterface> = state.competitions.filter((competition: CompetitionInterface) => {
        return competition.status === CompetitionConstants.CLOSED;
      });

      query = query.map((competition: CompetitionInterface) => {
        return Object.assign(competition, {
          buyerName: competition.theBuyer?.name ?? '',
          totalSuccessfulBids: competition.successfulBids.length,
          bidVolume: competition.successfulBids.reduce((previous: number, bid: BidsInterface) => previous += bid.offered_capacity, 0)
        });
      });

      return query;
    },
    getTopBuyers: (state: DataState): Array<TopBuyerInterface> => {
      return state.buyers.map((buyer: BuyerInterface) => {
        return {
          ...buyer,
          totalBoughtCompetitions: state.competitions
            .filter((competition: CompetitionInterface) => competition?.theBuyer?.id === buyer.id)
            .length
        };
      })
        .sort((a: TopBuyerInterface, b: TopBuyerInterface) => a.totalBoughtCompetitions - b.totalBoughtCompetitions)
        .slice(0, 10)
        .reverse();
    },
  },
});
