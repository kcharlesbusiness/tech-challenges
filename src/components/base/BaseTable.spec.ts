import { describe, it, expect } from 'vitest';
import BaseTable, { BaseTableInterface } from './BaseTable.vue';
import { render, RenderResult } from '@testing-library/vue';
import { createPinia, setActivePinia } from 'pinia';
import { useDataStore } from '../../store/dataStore';

describe('COMPONENT - BaseTable', () => {
  let wrapper: RenderResult;
  let dataStore: any;
  beforeEach(() => {
    setActivePinia(createPinia());
    dataStore = useDataStore();
  });

  it('renders a vue component without searchable input', () => {
    wrapper = render(BaseTable, {
      // @ts-ignore
      shallow: true,
      props: {
        title: 'BaseTable Title',
        data: dataStore.bids,
        searchable: false,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
  it('renders a vue component with searchable input', () => {
    wrapper = render(BaseTable, {
      // @ts-ignore
      shallow: true,
      props: {
        title: 'BaseTable Title',
        data: dataStore.bids,
        searchable: true,
      },
      // @ts-ignore
      stubs: ['o-table']
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
