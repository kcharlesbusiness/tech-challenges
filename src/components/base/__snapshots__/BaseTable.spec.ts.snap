// Vitest Snapshot v1

exports[`COMPONENT - BaseTable > renders a vue component with searchable input 1`] = `
"<section class=\\"v-base-table\\">
  <div class=\\"container\\">
    <div class=\\"v-base-table__table\\">
      <h2 class=\\"\\">BaseTable Title</h2>
      <section>
        <div><input type=\\"search\\" name=\\"search\\" placeholder=\\"search...\\" /></div>
        <!--v-if-->
      </section>
    </div>
  </div>
</section>"
`;

exports[`COMPONENT - BaseTable > renders a vue component without searchable input 1`] = `
"<section class=\\"v-base-table\\">
  <div class=\\"container\\">
    <div class=\\"v-base-table__table\\">
      <h2 class=\\"\\">BaseTable Title</h2>
      <section>
        <!--v-if-->
        <!--v-if-->
      </section>
    </div>
  </div>
</section>"
`;
